#!/usr/bin/env python
PACKAGE = "CORC"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("friction_ratio", double_t, 0, "Friction ratio", 0.5, 0, 1) # gen.add("friction_ratio", double_t, 0, "Friction ratio", 0.6, 0, 1)
gen.add("weight_ratio", double_t, 0, "Weight ratio", 0.5, 0, 1) # gen.add("weight_ratio", double_t, 0, "Weight ratio", 0.6, 0, 1)
gen.add("kp", double_t, 0, "P", 1.5, 0, 12) # gen.add("kp", double_t, 0, "P", 1.08, 0, 12)
gen.add("ki", double_t, 0, "I", 0.17, 0, 2)
gen.add("kd", double_t, 0, "D", 0.0, 0, 1)
gen.add("vel_thresh", double_t, 0, "vel_thresh", 15, 0, 50) # gen.add("vel_thresh", double_t, 0, "vel_thresh", 6, 0, 50)
gen.add("tau_thresh", double_t, 0, "tau_thresh", 0.2, 0, 3)
gen.add("sensor_cutoff_freq", double_t, 0, "sensor_cutoff_freq", 15, 0, 100)
gen.add("motor_torque_cutoff_freq", double_t, 0, "motor_torque_cutoff_freq", 50, 0, 100)
gen.add("tick_max", double_t, 0, "Reset interval (s)", 60, 0, 600)
gen.add("f_s_hys", double_t, 0, "Static friction hysteresis", 3, 0, 10)
gen.add("f_d_hys", double_t, 0, "Viscous friction hysteresis", 1.5, 0, 5)
gen.add("kp_mod", double_t, 0, "P mod", 0, 0, 6) # gen.add("kp_mod", double_t, 0, "P mod", 0, 0, 6)

gen.add("set_arom", bool_t, 1, "Search for min/max angle", False)
gen.add("set_prom", bool_t, 1, "Search for min/max angle", False)
gen.add("set_mvc", bool_t, 1, "Search for min/max torque", False)
gen.add("set_offset", bool_t, 1, "Update baseline torque", False)

gen.add("stim_calibrate", bool_t, 1, "Enable stimulation always ON", False)
gen.add("stim_amp_df", double_t, 0, "stim_amp_df", 0, 0, 25)
gen.add("stim_amp_pf", double_t, 0, "stim_amp_pf", 0, 0, 25)

interaction_enum = gen.enum([
                   gen.const("zero_calibration", int_t, 0, "controller 1"),
                   gen.const("zero_torque", int_t, 1, "controller 1"),
                   gen.const("follow_position", int_t, 2, "controller 2"),
                   gen.const("follow_torque", int_t, 3, "controller 3"),
                   gen.const("virtual_spring", int_t, 4, "controller 4"),
                   gen.const("transparency", int_t, 5, "controller 5"),
                   gen.const("step_angle", int_t, 6, "controller 6"),
                   gen.const("center_angle", int_t, 7, "controller 7"),
                   gen.const("passive_rom", int_t, 8, "controller 8"),
                   gen.const("system_id", int_t, 10, "controller 10"),
                   gen.const("send_high", int_t, 11, "controller 11"),
                   gen.const("matching_reference", int_t, 12, "controller 12")],
                    "MultiState Controller Mode")

gen.add("controller_mode", int_t, 1, "Controller Mode", 1, 0, 12, edit_method=interaction_enum) # 11 is the maximum value

exit(gen.generate(PACKAGE, "CORC", "dynamic_params"))